ledger/simulation/tracer.go

```diff
@@ -267,10 +299,51 @@ func (tracer *evalTracer) BeforeOpcode(cx *logic.EvalContext) {
 			txnTrace = tracer.execTraceStack[len(tracer.execTraceStack)-1]
 		}
 		*txnTrace.programTraceRef = append(*txnTrace.programTraceRef, tracer.makeOpcodeTraceUnit(cx))
+
+		if tracer.result.ReturnStackChange() {
+			latestOpcodeTraceUnit := &(*txnTrace.programTraceRef)[len(*txnTrace.programTraceRef)-1]
+			latestOpcodeTraceUnit.appendDeletedStackValue(cx, tracer)
+			latestOpcodeTraceUnit.DisassembledLine, _ = cx.DisassembleLine()
+		}
+	}
+}
+
+func (o *OpcodeTraceUnit) appendAddedStackValue(cx *logic.EvalContext, tracer *evalTracer) {
+	// keep adding stuffs to added slice
+	if tracer.stackChangeExplanation.Additions == 0 {
+		return
+	}
+	stackCopy := cx.Stack()
+	for i := tracer.stackHeightAfterDeletion; i < len(stackCopy); i++ {
+		if stackCopy[i].TEALType() == basics.TealBytesType {
+			o.Added = append(o.Added, TealValue{
+				Type:  basics.TealBytesType,
+				Bytes: string(stackCopy[i].Bytes),
+			})
+		} else {
+			o.Added = append(o.Added, TealValue{
+				Type: basics.TealUintType,
+				Uint: stackCopy[i].Uint,
+			})
```

Comment:
similar simplification from `basics.TealType`

and, I think if these places are the only place it's used, the new `TEALType()` method can be deleted.